File: problems/problems_560/Solution.cpp
//go:build ignore
#include "cpp/common/Solution.h"
#include <unordered_map>


using namespace std;
using json = nlohmann::json;

class Solution {
public:
    int subarraySum(vector<int>& nums, int k) {
        int ans = 0, s = 0;
        unordered_map<int, int> counter{{0, 1}}; // s[0]=0 单独统计
        for (auto num: nums) {
            s += num;
            ans += counter.find(s - k) != counter.end() ? counter[s - k] : 0;
            counter[s]++;
        }
        return ans;
    }
};

json leetcode::qubh::Solve(string input_json_values) {
	vector<string> inputArray;
	size_t pos = input_json_values.find('\n');
	while (pos != string::npos) {
		inputArray.push_back(input_json_values.substr(0, pos));
		input_json_values = input_json_values.substr(pos + 1);
		pos = input_json_values.find('\n');
	}
	inputArray.push_back(input_json_values);

	Solution solution;
	vector<int> nums = json::parse(inputArray.at(0));
	int k = json::parse(inputArray.at(1));
	return solution.subarraySum(nums, k);
}

File: problems/problems_560/solution.py
import solution
from typing import *
from collections import defaultdict


class Solution(solution.Solution):
    def solve(self, test_input=None):
        return self.subarraySum(*test_input)

    def subarraySum(self, nums: List[int], k: int) -> int:
        ans = s = 0
        counter = defaultdict(int)
        counter[0] = 1
        for num in nums:
            s += num
            ans += counter[s - k]
            counter[s] += 1
        return ans
